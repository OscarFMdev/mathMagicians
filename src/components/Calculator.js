import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';

const Buttons = ({ myFunction }) => (
  <>
    <button onClick={myFunction} type="button" className="blue-button">AC</button>
    <button onClick={myFunction} type="button" className="blue-button">+/-</button>
    <button onClick={myFunction} type="button" className="blue-button">%</button>
    <button onClick={myFunction} type="button" className="orange-button">รท</button>
    <button onClick={myFunction} type="button" className="blue-button">7</button>
    <button onClick={myFunction} type="button" className="blue-button">8</button>
    <button onClick={myFunction} type="button" className="blue-button">9</button>
    <button onClick={myFunction} type="button" className="orange-button">x</button>
    <button onClick={myFunction} type="button" className="blue-button">4</button>
    <button onClick={myFunction} type="button" className="blue-button">5</button>
    <button onClick={myFunction} type="button" className="blue-button">6</button>
    <button onClick={myFunction} type="button" className="orange-button">-</button>
    <button onClick={myFunction} type="button" className="blue-button">1</button>
    <button onClick={myFunction} type="button" className="blue-button">2</button>
    <button onClick={myFunction} type="button" className="blue-button">3</button>
    <button onClick={myFunction} type="button" className="orange-button">+</button>
    <button onClick={myFunction} type="button" className="blue-button large">0</button>
    <button onClick={myFunction} type="button" className="blue-button">.</button>
    <button onClick={myFunction} type="button" className="orange-button">=</button>
  </>
);

const Calculator = () => {
  const [display, setDisplay] = useState(0);
  useEffect(() => {
    setDisplay({
      total: 0,
      next: null,
      operation: null,
    });
  }, []);

  const handleClick = ({ target }) => {
    setDisplay(calculate(display, target.textContent));
  };

  const { total, next, operation } = display;
  return (
    <>
      <section className="calculator-container">
        <div className="calc-quote">Let&apos;s do dome Math!</div>
        <div className="calculator">
          <div className="result-box">
            <p>
              {total}
              {operation}
              {next}
            </p>
          </div>
          <ul className="operation-grid">
            <Buttons myFunction={handleClick} />
          </ul>
        </div>
      </section>
    </>
  );
};

Buttons.propTypes = {
  myFunction: PropTypes.func.isRequired,
};

export default Calculator;
