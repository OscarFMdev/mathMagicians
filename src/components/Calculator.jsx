import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';

const Buttons = ({ myFunction }) => (
  <>
    <button onClick={myFunction} type="button" className="gray-button">AC</button>
    <button onClick={myFunction} type="button" className="gray-button">+/-</button>
    <button onClick={myFunction} type="button" className="gray-button">%</button>
    <button onClick={myFunction} type="button" className="orange-button">รท</button>
    <button onClick={myFunction} type="button" className="gray-button">7</button>
    <button onClick={myFunction} type="button" className="gray-button">8</button>
    <button onClick={myFunction} type="button" className="gray-button">9</button>
    <button onClick={myFunction} type="button" className="orange-button">x</button>
    <button onClick={myFunction} type="button" className="gray-button">4</button>
    <button onClick={myFunction} type="button" className="gray-button">5</button>
    <button onClick={myFunction} type="button" className="gray-button">6</button>
    <button onClick={myFunction} type="button" className="orange-button">-</button>
    <button onClick={myFunction} type="button" className="gray-button">1</button>
    <button onClick={myFunction} type="button" className="gray-button">2</button>
    <button onClick={myFunction} type="button" className="gray-button">3</button>
    <button onClick={myFunction} type="button" className="orange-button">+</button>
    <button onClick={myFunction} type="button" className="gray-button large">0</button>
    <button onClick={myFunction} type="button" className="gray-button">.</button>
    <button onClick={myFunction} type="button" className="orange-button">=</button>
  </>
);

const Calculator = () => {
  const [display, setDisplay] = useState(0);
  useEffect(() => {
    setDisplay({
      total: 0,
      next: null,
      operation: null,
    });
  }, []);

  const handleClick = (e) => {
    setDisplay(calculate(display, e.target.textContent));
  };

  const { total, next, operation } = display;
  return (
    <section className="calculator-container">
      <div className="calculator">
        <div className="gray-box">
          <p>
            {total}
            {operation}
            {next}
          </p>
        </div>
        <ul className="operation-grid">
          <Buttons myFunction={handleClick} greet="Hello" />
        </ul>
      </div>
    </section>
  );
};

Buttons.propTypes = {
  myFunction: PropTypes.func.isRequired,
};

export default Calculator;
