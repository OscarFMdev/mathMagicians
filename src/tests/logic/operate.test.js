import { fireEvent, render, screen } from '@testing-library/react';
import Calculator from '../../components/Calculator';
import operate from '../../logic/operate';

describe('Operate tests', () => {
  /* OPERATION TESTS */
  test('2 x 5 should return \'10\'', () => {
    /* Arrange */
    const numberOne = 2;
    const numberTwo = 5;
    const operation = 'x';

    /* Act */
    const result = operate(numberOne, numberTwo, operation);

    /* Assert */
    expect(result).toBe('10');
  });

  /* RENDERING TESTS */
  test('0 + 0 should render 0', () => {
    /* Arrange */
    const numberOne = 0;
    const numberTwo = 0;
    const operation = '+';

    /* Act */
    const { container } = render(operate(numberOne, numberTwo, operation));

    /* Assert */
    expect(container).toMatchSnapshot();
  });

  test('5 x 7 should render a div containing 35', () => {
    const numberOne = 5;
    const operation = 'x';
    const numberTwo = 7;

    const result = '35';
    render(operate(numberOne, numberTwo, operation));

    /* Assert */
    expect(screen.getByText(result)).toBeDefined();
  });

  /* USER INTERACTION */
  test('should modify the result-box content after user pressing the buttons "7"', () => {
    /* Arrange */
    const { container } = render(<Calculator />);
    const buttons = container.querySelectorAll('button');
    const buttonSeven = buttons[4];

    /* Act */
    fireEvent.click(buttonSeven);
    const resultBox = container.querySelector('.result-box > p').innerHTML;

    /* Assert */
    expect(resultBox).toBe('7');
  });
});
