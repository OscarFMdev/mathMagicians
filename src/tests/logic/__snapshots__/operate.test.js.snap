// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operate tests 0 + 0 should render 0 1`] = `
<div>
  0
</div>
`;
